### PR TITLE
TestPlugin cache refactored.

### PR DESCRIPTION
**Pull request by harshjain2**
1. Refactored Test Plugin Cache to support loading for individual extensions.

### CODE REVIEW

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginCache.cs [hunk 1]
```diff
@@ -113,7 +104,7 @@ public IEnumerable<string> PathToAdditionalExtensions
         {
```
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginCache.cs]: Should we rename it to `PathToExtensions`

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs [hunk 2]
```diff
@@ -23,45 +23,96 @@ public class TestExtensions
         internal Dictionary<string, TestDiscovererPluginInformation> TestDiscoverers { get; set; }
 
         /// <summary>
+        /// Gets or sets a value indicating whether are test discoverers cached.
+        /// </summary>
+        internal bool AreTestDiscoverersCached { get; set; }
```
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs]: Where are we setting the value of variable `AreTestDiscoverersCached`

File: src/Microsoft.TestPlatform.Common/Logging/TestLoggerExtensionManager.cs [hunk 3]
```diff
@@ -53,12 +54,12 @@ public static TestLoggerExtensionManager Create(IMessageLogger messageLogger)
             IEnumerable<LazyExtension<ITestLogger, ITestLoggerCapabilities>> filteredTestExtensions;
             IEnumerable<LazyExtension<ITestLogger, Dictionary<string, object>>> unfilteredTestExtensions;
 
-            TestPluginManager.Instance.GetTestExtensions<ITestLogger, ITestLoggerCapabilities, TestLoggerMetadata>(
+            TestPluginManager.Instance.GetSpecificTestExtensions<TestLoggerPluginInformation, ITestLogger, ITestLoggerCapabilities, TestLoggerMetadata>(
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.Common/Logging/TestLoggerExtensionManager.cs]: >TestLoggerPluginInformation [](start = 65, length = 27)

Would we ever call TestLoggerPluginInformation with a different regex pattern? Maybe the regex patter should be in TestPluginInfo itself. Dont see a reason for it to be propagated everywhere when TestPluginInfo is anyway passed through.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs [hunk 4]
```diff
@@ -119,11 +119,7 @@ public override void Close()
         private void InitializeExtensions(IEnumerable<string> sources)
         {
             var sourceList = sources.ToList();
-            var extensions = this.testHostManager.GetTestPlatformExtensions(sourceList).ToList();
-            if (TestPluginCache.Instance.PathToAdditionalExtensions != null)
-            {
-                extensions.AddRange(TestPluginCache.Instance.PathToAdditionalExtensions);
-            }
+            var extensions = this.testHostManager.GetTestPlatformExtensions(sourceList, TestPluginCache.Instance.PathToAdditionalExtensions).ToList();
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs]: >.ToList(); [](start = 140, length = 10)

nit: use IEnumerable.

File: src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs [hunk 5]
```diff
@@ -100,7 +100,7 @@ public DotnetTestHostManager()
         /// <remarks>
         /// Dependency resolution for .net core projects are pivoted by the test project. Hence each test
         /// project must be launched in a separate test host process.
-        /// </remarks>
+        /// </remarks>InitializeExtensions
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs]: >InitializeExtensions [](start = 22, length = 20)

cleanup?

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs [hunk 6]
```diff
@@ -23,45 +23,96 @@ public class TestExtensions
...
+        /// <typeparam name="TPluginInfo">
+        /// Type of plugin info.
+        /// </typeparam>
+        /// <param name="newExtensions">
+        /// The info about new extensions discovered
+        /// </param>
+        /// <returns>
+        /// The <see cref="Dictionary"/> of extensions discovered
+        /// </returns>
+        internal Dictionary<string, TPluginInfo> AddExtension<TPluginInfo>(
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs]: >AddExtension [](start = 49, length = 12)

misleading name: It says Add but we are setting it. Is that intended?

**Comment by AbhitejJohn** [comment]
>     public class TestExtensions

What exactly is the responsibility of this class?

---
Refers to: src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs:16 in b0dc0f3. [](commit_id = b0dc0f39b6cb3fad4d37986979eec8e90ebcf81d, deletion_comment = False)

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginCache.cs [hunk 7]
```diff
@@ -408,16 +331,69 @@ internal virtual string[] GetFilesInDirectory(string path, string searchPattern)
...
-            return discoverer.GetTestExtensionsInformation(extensionPaths, this.loadOnlyWellKnownExtensions);
+            var pluginInfos = discoverer.GetTestExtensionsInformation<TPluginInfo, TExtension>(extensionPaths, this.loadOnlyWellKnownExtensions);
+
+            // Add extensions discovered to the cache.
+            if (this.TestExtensions == null)
+            {
+                this.TestExtensions = new TestExtensions();
+            }
+
+            this.TestExtensions.AddExtension<TPluginInfo>(pluginInfos);
```
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginCache.cs]: Should we update our central cache?

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginCache.cs [hunk 8]
```diff
@@ -408,16 +331,69 @@ internal virtual string[] GetFilesInDirectory(string path, string searchPattern)
...
+            // Check if extensions from this assembly have already been discovered.
+            var extensions = this.TestExtensions?.GetExtensionsDiscoveredFromAssembly<TPluginInfo>(this.TestExtensions.GetTestExtensionCache<TPluginInfo>(), extensionAssembly);
+
+            if (extensions != null)
+            {
+                return extensions;
+            }
+
+            this.GetTestExtensions<TPluginInfo, TExtension>(new List<string>() { extensionAssembly });
+            return this.TestExtensions?.GetExtensionsDiscoveredFromAssembly<TPluginInfo>(this.TestExtensions.GetTestExtensionCache<TPluginInfo>(), extensionAssembly);
```
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginCache.cs]: I don't think we need this line.

File: src/vstest.console/Processors/TestAdapterPathArgumentProcessor.cs [hunk 9]
```diff
@@ -204,7 +214,7 @@ public ArgumentProcessorResult Execute()
         /// <returns> The list of test adapter and logger assemblies. </returns>
         internal virtual IEnumerable<string> GetTestAdaptersFromDirectory(string directory)
```
Comment (Faizan2304) [inline src/vstest.console/Processors/TestAdapterPathArgumentProcessor.cs]: Do we need this function now? I think we can remove it and we should add logic to find extension in testplatform.UpdateTestAdapterPaths( string runSettings )

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginCache.cs (hunk 1 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/TestPluginCache.cs]: Resolved.

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs (hunk 2 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs]: Resovled.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs (hunk 4 above)
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs]: resolved.

**Review by Faizan2304 — APPROVED** [review]
👍 LGTM

Please add following scenario in your test matrix:
1) Run cpp tests
2) Pass logger through /TestAdapterpath

File: src/Microsoft.TestPlatform.Common/Logging/TestLoggerExtensionManager.cs (hunk 3 above)
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.Common/Logging/TestLoggerExtensionManager.cs]: 👍

File: src/vstest.console/Processors/TestAdapterPathArgumentProcessor.cs [hunk 10]
```diff
@@ -153,33 +155,58 @@ public void Initialize(string argument)
...
+            var fullPathTestAdapterPaths = new List<string>();
 
             try
             {
                 // Remove leading and trailing ' " ' chars...
                 argument = argument.Trim().Trim(new char[] { '\"' });
 
-                customAdaptersPath = Path.GetFullPath(argument);
-                if (!fileHelper.DirectoryExists(customAdaptersPath))
+                var testadapterPaths = argument.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries).ToList();
```
Comment (Faizan2304) [inline src/vstest.console/Processors/TestAdapterPathArgumentProcessor.cs]: TestAdapterPath can not have more than one path.

File: src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs [hunk 11]
```diff
@@ -23,45 +23,96 @@ public class TestExtensions
...
+
+        /// <summary>
         /// Gets or sets test setting provider extensions.
         /// </summary>
         internal Dictionary<string, TestSettingsProviderPluginInformation> TestSettingsProviders { get; set; }
 
         /// <summary>
+        /// Gets or sets a value indicating whether are test settings providers cached.
+        /// </summary>
+        internal bool AreTestSettingsProvidersCached { get; set; }
```
Comment (AbhitejJohn) [inline src/Microsoft.TestPlatform.Common/ExtensionFramework/Utilities/TestExtensions.cs]: Are we planning to move these out to their respective plugin managers? This is becoming a God class.

**Comment by AbhitejJohn** [comment]
Please file an issue for the two broad level issues that came out as part of this PR:
1. The regex pattern that is passed all over which isn't needed.
2. The God level TestExtensions.cs class that needs to transfer responsibility over to each of the plugin managers.

File: src/vstest.console/Processors/TestAdapterPathArgumentProcessor.cs (hunk 10 above)
Comment (harshjain2) [inline src/vstest.console/Processors/TestAdapterPathArgumentProcessor.cs]: Reverted to original behaviour.

File: src/vstest.console/Processors/TestAdapterPathArgumentProcessor.cs [hunk 12]
```diff
@@ -155,63 +155,55 @@ public void Initialize(string argument)
...
 
-            this.commandLineOptions.TestAdapterPath = fullPathTestAdapterPathsString;
-
-            if (fullPathTestAdapterPaths.Count <= 0)
-            {
-                // Print a warning about not finding any test adapter in provided path...
-                this.output.Warning(CommandLineResources.NoAdaptersFoundInTestAdapterPath, argument);
-                this.output.WriteLine(string.Empty, OutputLevel.Information);
-            }
+            this.commandLineOptions.TestAdapterPath = customAdaptersPath;
```
Comment (Faizan2304) [inline src/vstest.console/Processors/TestAdapterPathArgumentProcessor.cs]: Can we please check uses of this. earlier it were taking single path but now u are passing list of paths

**Comment by harshjain2** [comment]
@dotnet-bot test this please

**Comment by codito** [comment]
@dotnet-bot test this please

**Pull request merged**